content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+---
+title: callPackage, a tool for the lazy
+date: 2022-08-25
+extra:
+  author: Norbert Melzer
+---
+
+In [`nixpkgs`](https://github.com/nixos/nixpkgs), there is a massive use of the
+`callPackage` function, which provides us with a lot of benefits.
+
+## Basic examples
+
+Though before even discussing the benefits, lets see how it actually gets used.
+
+Given are these 2 files `hello.nix` and `default.nix`:
+
+```nix
+# Usually I would prefer the usage of nix flakes, though that would
+# introduce a lot of boilerplate in the examples for no reason,
+# therefore I will use diamond paths and `import` for simplicity.
+```
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> {}; in
+pkgs.callPackage ./hello.nix {}
+```
+
+```nix
+# hello.nix
+{writeShellScriptBin}:
+writeShellScriptBin "hello" ''
+  echo "hello, world!"
+''
+```
+
+Building using `nix-build` will produce `./result/bin/hello`, and running it will
+nicely greet you.
+
+As you can see, `writeShellScriptBin` gets passed in by `callPackage` automatically.
+
+For this simple setup, having to create an extra file, seems to be a lot of
+boilerplate, though if you continue reading, you will see, it is worth it!
+
+## 1. Benefit: parametrized builds
+
+Now lets change the `default.nix`:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+}
+```
+
+Now we build using `nix-build -A hello`, the outcome will be the same as above.
+
+Now to "parametrize" the build we also change the `hello.nix` a bit:
+
+```nix
+# hello.nix
+{ writeShellScriptBin
+, audience ? "world"
+}:
+writeShellScriptBin "hello" ''
+  echo "hello, ${audience}!"
+''
+```
+
+Building this will still yield the same output as before, though now things get
+interesting, alter your `default.nix` yet another time:
```

Comment:
```suggestion
Building this will still yield the same output as before.

Things get more interesting when we alter `default.nix` another time, by adding a new attribute `people` to the result.

Note how we pass a parameter `audience` in the second argument to `callPackage`, which is handed over to the function defined in `hello.nix`:
```